src/Features/CSharp/Portable/CodeCleanup/CSharpCodeCleanupService.cs

```diff
@@ -0,0 +1,190 @@
+
+            // do the remove usings after code fix, as code fix might remove some code which can results in unused usings.
+            document = await RemoveSortUsingsAsync(document, docOptions, cancellationToken).ConfigureAwait(false);
+
+            return await Formatter.FormatAsync(document).ConfigureAwait(false);
+        }
+
+        private async Task<Document> RemoveSortUsingsAsync(Document document, DocumentOptionSet docOptions, CancellationToken cancellationToken)
+        {
+            // remove usings
+            if (docOptions.GetOption(CodeCleanupOptions.RemoveUnusedImports))
+            {
+                var removeUsingsService = document.GetLanguageService<IRemoveUnnecessaryImportsService>();
+                if (removeUsingsService != null)
+                {
+                    document = await removeUsingsService.RemoveUnnecessaryImportsAsync(document, cancellationToken).ConfigureAwait(false);
+                }
+            }
+
+            // sort usings
+            if (docOptions.GetOption(CodeCleanupOptions.SortImports))
+            {
+                document = await OrganizeImportsService.OrganizeImportsAsync(document, cancellationToken).ConfigureAwait(false);
+            }
+
+            return document;
+        }
+
+        private async Task<Document> ApplyCodeFixesAsync(Document document, DocumentOptionSet docOptions, CancellationToken cancellationToken)
+        {
+            var fixAllService = document.Project.Solution.Workspace.Services.GetService<IFixAllGetFixesService>();
+
+            var dummy = new ProgressTracker();
+            foreach (var diagnosticId in GetEnabledDiagnosticIds(docOptions))
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var syntaxTree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+                var textSpan = new TextSpan(0, syntaxTree.Length);
+
+                var fixCollection = await _codeFixService.GetFixesAsync(document, textSpan, diagnosticId, cancellationToken).ConfigureAwait(false);
+                if (fixCollection == null)
+                {
+                    continue;
+                }
+
+                var fixAll = fixCollection.FixAllState;
+                var solution = await fixAllService.GetFixAllChangedSolutionAsync(
+                    fixAll.CreateFixAllContext(dummy, cancellationToken)).ConfigureAwait(false);
+
+                document = solution.GetDocument(document.Id);
+            }
+
+            return document;
+        }
+
+        private List<string> GetEnabledDiagnosticIds(DocumentOptionSet docOptions)
+        {
+            var diagnosticIds = new List<string>();
+
+            foreach (var codeCleanupOptions in _optionDiagnosticsMappings.Keys)
+            {
+                if (docOptions.GetOption(codeCleanupOptions))
+                {
+                    diagnosticIds.AddRange(_optionDiagnosticsMappings[codeCleanupOptions]);
```

Comment:
kv.Value

Reply:
by the way, is ordering matter? dictionary doesn't maintain ordering of its entries? you might want to just use ImmutableArray((option option, Immutablearray<string> ids)) rather than dictionary since it looks like you won't actually need indexer access for your dictionary.